File: Data-Structures-and-Algorithms/0347-TopKFrequentElements/topKFrequent.js
const topKFrequent = function(nums, k) {
    // because k is in the range of 1 to number of unique numbers
    if (nums.length == k) {
        return nums
    }
        
    // create a map that records the number as key and count as value (On and On)
    const map = new Map();
    for (const n of nums) {
        if (map[n] == null) {
            map[n] = 1
        } else {
            map[n] ++
        }
    }

    // create a bucketlist with count as the index (index 0 contains numbers that has been repeated once, and index 1....)
    let bucketlist = []
    for (const n of nums) {
        bucketlist.push([])
    }
    for (const [key, value] of Object.entries(map)) {
        bucketlist[value - 1].push(parseInt(key))
    }

    // create a final list that contains all entries of numbers from least to most appeared
    let finalist = []
    for (bucket of bucketlist) {
        finalist.push(...bucket)
    }

    return finalist.slice(-k)
};

var nums = [1, 1, 1, 1, 4, 5, 6, 7, 5, 6, 8, 1, 4, 5, 0, 1, 2, 5, 0, 5, 5, 5, 5, 5, 5, 5, 6]
var k = 3
console.log(topKFrequent(nums, k))

// O(N) for time, because we traverse through the iteration for multiple times but never nested within each other
// O(N) for space for similar reasons
// Top K numbers with the most appearences

File: Data-Structures-and-Algorithms/0347-TopKFrequentElements/topKFrequent.py
# i need the dict function from collections and I am going to use bucketlist
import collections
class Solution(object):
    def topKFrequent(self, nums, k):
        # we cannot sort for sorting violates the complexity
        # initiate a dictionary with integer type 
        myhash = collections.defaultdict(int)

        # because k is in the range of 1 to number of unique numbers
        if len(nums) == k:
            return nums

        # add nums and their counts into the dict
        for n in nums:
            myhash[n] += 1

        # initiate a bucketlist that uses index as count (elements in the 0th list has a count of 1)
        bucketlist = [[] for i in range(len(nums))]
        for number, count in myhash.items():
            bucketlist[count-1].append(number)
        print(bucketlist)

        # put everything into a stack, and the last bucket which has the highest counts for numbers therein will be at the end of the stack...
        finalist = []
        for bucket in bucketlist:
            finalist.extend(bucket)
        print(finalist)

        # return the last k elements in the list
        res = finalist[-k:]

        return res   

nums = [1, 1, 1, 1, 4, 5, 6, 7, 5, 6, 8, 1, 4, 5, 0, 1, 2, 5, 0, 5, 5, 5, 5, 5, 5, 5, 6]
k = 3
print(Solution().topKFrequent(nums, k))

# O(N) for time, because we traverse through the iteration for multiple times but never nested within each other
# O(N) for space for similar reasons
# Top K numbers with the most appearences
